Use gRPC to communicate data from app host to dashboard

drewnoakes opened the pull request:
Includes PR #1415, with feedback from that PR addressed here.

This is another step towards #1003. Ultimately the dashboard will run in a separate process to the app host. That will require a network connection between the processes for data.

This PR splits the front and back end to use separate types, with a gRPC channel between them.

The `.proto` defined in #1274 is used here, with some modifications following the integration.

Contributes to #436.


 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1476)

drewnoakes committed:
Model per-type resource data as extension data

The UI will have a single object type for all resources, treating the uniformly. That object has both a fixed set of fields, and now its own extensible metadata.

In the protocol between ResourceService and Dashboard, we'll need an extensible way to model these resource data. This allows the UI to work with resource types it hasn't seen before.

drewnoakes committed:
Split snapshots from view models

We are moving towards removing `Aspire.Hosting`'s reference to `Aspire.Dashboard`. In this new model we will have different types on the back end and front end.

- DCP has its own types (e.g. `CustomResource`), which remain untouched, and exist only on the back end.
- `ResourceSnapshot` is a new abstract base class for `ContainerSnapshot`, `ExecutableSnapshot` and `ProjectSnapshot`. These are immutable, have no view-specific data, and will only be used on the back end. In the near future, these objects will be used to populate gRPC messages.
- `ResourceViewModel` remains, though only on the front end. It no longer has subclasses. We are designing the dashboard to support arbitrary resource types. In the near future, these objects will be constructed and updated in response to gRPC messages.

As the front end no longer has subclasses of `ResourceViewModel`, some of the hard-coded resource-specific logic has been changed to work on the additional data within a resource.

We also introduce a `KnownResourceTypes` static class with consts for known resource type names (e.g. "Container").

drewnoakes committed:
Remove obsolete TODOs

drewnoakes committed:
Fix tests

drewnoakes committed:
Merge branch 'main' into split-dto

drewnoakes committed:
WIP

drewnoakes committed:
Rename ResourceDataKeys to KnownProperties

drewnoakes committed:
WIP

drewnoakes committed:
Reorder members

drewnoakes committed:
Move types to their own source files

drewnoakes committed:
Split endpoint snapshot from view model

drewnoakes committed:
Consistent naming in razor files

drewnoakes committed:
Add ctor to EnvironmentVariableViewModel

drewnoakes committed:
Complete ResourceViewModel construction from Resource message

drewnoakes committed:
Make types internal

drewnoakes committed:
Add words to dictionary

drewnoakes committed:
Use KnownResourceTypes

drewnoakes committed:
Pass correct type to logger factory

drewnoakes committed:
Rename ResourceService to DashboardClient

drewnoakes committed:
Refactor

drewnoakes committed:
Dashboard receives data via gRPC

Uses the gRPC `.proto` file added in a previous commit (with some small changes) to model data sent between the back end (app host) and front end (dashboard). Currently this occurs within the same process, but future work will split the dashboard out of the app host.

- Object flow is now "DCP -> snapshots -> gRPC messages -> view models".
- Move `DashboardWebApplication` initialization out of `DcpHostService` and into the new `DashboardWebApplicationHost` in preparation for future split.
- Split `ResourceSubscription`/`ResourceChange`/`ResourceChangeType` into both snapshot and view model versions, for use on back/front ends respectively.

drewnoakes committed:
Merge branch 'main' into add-grpc

drewnoakes committed:
Fix ports display in resource grid

drewnoakes committed:
Use extension methods to check resource type

drewnoakes committed:
Remove obsolete TODO

drewnoakes committed:
Document use of Math.Round

drewnoakes committed:
Consolidate types

drewnoakes committed:
Rename comparer

This was missed in a previous commit where we renamed ResourceDataKeys to KnownProperties.

drewnoakes committed:
Remove ResourceId

We don't use this consistently throughout. Plenty of places still require a unique resource name as a key. If we want to change this later, we should do it wholesale across the system in future work.

JamesNK commented on src/Aspire.Hosting/Dashboard/DashboardServiceData.cs:
```diff
@@ -0,0 +1,57 @@
+
+        return sequence is null ? null : Enumerate();
+
+        async IAsyncEnumerable<IReadOnlyList<(string Content, bool IsErrorMessage)>> Enumerate()
```
nit: I prefer static local methods. Could pass in the cts and sequence as parameters and make this static.

drewnoakes replied:
I use both static and non-static local methods on a case-by-case basis.

JamesNK commented on src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:
```diff
@@ -0,0 +1,114 @@
+                    // which cannot negotiate between HTTP/1.1 and HTTP/2.
+                    options.Protocols = HttpProtocols.Http2;
+
+                    if (string.Equals(uri.Scheme, "https", StringComparison.Ordinal))
```
Ignore case? Someone could pass in uppercase strings in the env variable.

(or does Uri normalize scheme to lowercase?)

drewnoakes replied:
This code exists in a few places today, and I copied it from somewhere. Please file a separate issue if it is actually an issue.

JamesNK commented on src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:
```diff
@@ -0,0 +1,114 @@
+            return;
+        }
+
+        var builder = WebApplication.CreateBuilder();
```
What about configuring logging for the web app?

I think the app should use the same logging as its container. I'm not sure the best way to do that off the top of my head. Maybe pass a logger factory to the DashboardServiceHost constructor and register it as a singleton?

JamesNK replied:
How was this resolved?

drewnoakes replied:
Today the dashboard continues to use the app host's logging. Configuring the dashboard's logging can be done when we pull it out into a standalone tool. @mitchdenny and I have been looking into that for a future PR.

JamesNK commented on src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:
```diff
@@ -0,0 +1,114 @@
+        _app = builder.Build();
+
+        _app.MapGrpcService<DashboardService>();
+        _app.MapGet("/", () => "Communication with gRPC endpoints must be made through a gRPC client. To learn how to create a client, visit: https://go.microsoft.com/fwlink/?linkid=2086909");
```
This can be removed. No one should ever be browsing to this port. And if they do, they're probably get an error about HTTP/2 on an unsecured port before they see this error.

drewnoakes replied:
Indeed. You'd see this, not the route:

![image](https://github.com/dotnet/aspire/assets/350947/77fa49b1-6325-44ae-86f2-017e17432cc0)

JamesNK commented on src/Aspire.Hosting/Dashboard/DashboardServiceHost.cs:
```diff
@@ -0,0 +1,114 @@
+    {
+        if (!options.DashboardEnabled)
+        {
+            return;
```
Log that it isn't starting

JamesNK commented on src/Aspire.Hosting/Dashboard/DashboardWebApplicationHost.cs:
```diff
@@ -0,0 +1,50 @@
+
+    async Task IHostedService.StartAsync(CancellationToken cancellationToken)
+    {
+        if (_dashboardWebApp is not null)
```
Logging for when it is null

drewnoakes replied:
Added to the constructor which has more context on the reason.

JamesNK commented on src/Aspire.Hosting/Dashboard/DashboardWebApplicationHost.cs:
```diff
@@ -0,0 +1,50 @@
+
+    async Task IHostedService.StopAsync(CancellationToken cancellationToken)
+    {
+        if (_dashboardWebApp is not null)
```
Logging for when it is null

JamesNK commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,309 @@
+                InitialState: _resourceByName.Values.ToImmutableArray(),
+                Subscription: StreamUpdates());
+
+            async IAsyncEnumerable<ResourceViewModelChange> StreamUpdates()
```
FYI this has the same problem as I commented on here: https://github.com/dotnet/aspire/pull/1348#discussion_r1428707324

Subscriptions won't be stopped when they're canceled.

drewnoakes replied:
I'm going to take a separate pass through all this and ensure nothing's running longer or more frequently than it should. I saw your comment there and will address it in a subsequent PR as the problem already exists on main and this PR is large enough as is.

JamesNK commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,309 @@
+
+        _ = Task.Run(() => ConnectAndStayConnectedAsync(address, _cts.Token), _cts.Token);
+
+        async Task ConnectAndStayConnectedAsync(Uri address, CancellationToken cancellationToken)
```
What you're doing here with recreating the channel and client isn't necessary. A channel will automatically connect and reconnect when required.

Retrying should focus on the call to the server. It is the what needs to be recreated on error.

drewnoakes replied:
Oh that's good to know. I'll take another look. There's more work I want to do around reconnects, so I might have a follow up for that.

JamesNK commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,309 @@
+    {
+        if (Interlocked.Exchange(ref _state, StateDisposed) is not StateDisposed)
+        {
+            await _cts.CancelAsync().ConfigureAwait(false);
```
Dispose should dispose the channel. That ensures current gRPC calls are stopped and TCP connections are closed.

drewnoakes committed:
Add API doc

drewnoakes committed:
Logging

drewnoakes committed:
Don't treat cancellations as errors

This prevents a cancellation

drewnoakes committed:
Join async connection management on dispose

drewnoakes commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,309 @@
+
+        _ = Task.Run(() => ConnectAndStayConnectedAsync(address, _cts.Token), _cts.Token);
+
+        async Task ConnectAndStayConnectedAsync(Uri address, CancellationToken cancellationToken)
```
I prefer having methods, just like values, in the narrowest necessary scope. It helps highlight where things are used, makes improvements easier to spot.

drewnoakes commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,309 @@
+                            errorCount = 0;
+                        }
+
+                        if (response.KindCase == WatchResourcesUpdate.KindOneofCase.InitialData)
```
I prefer if/else here. Switch is bulky.

drewnoakes commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,309 @@
+
+    public string ApplicationName => _applicationName ?? "";
+
+    ResourceViewModelSubscription IDashboardClient.SubscribeResources()
```
This will only be called via the interface. I prefer advertising when an interface method is implementing that interface in such cases.

drewnoakes commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,309 @@
+            {
+                var call = client.WatchResources(new WatchResourcesRequest { IsReconnect = errorCount != 0 }, cancellationToken: cancellationToken);
+
+                await foreach (var response in call.ResponseStream.ReadAllAsync(cancellationToken: cancellationToken))
```
Removing it results in CA2016. Eaiser just to add it here than to suppress the diagnostic.

drewnoakes commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,309 @@
+            new WatchResourceConsoleLogsRequest() { ResourceName = resourceName },
+            cancellationToken: cancellationToken);
+
+        await foreach (var response in call.ResponseStream.ReadAllAsync(cancellationToken: cancellationToken))
```
Again, removing it triggers an analyzer.

drewnoakes commented on src/Aspire.Hosting/Dashboard/DashboardService.cs:
```diff
@@ -0,0 +1,109 @@
+            {
+                var change = new WatchResourcesChange();
+
+                if (update.ChangeType is ResourceSnapshotChangeType.Upsert)
```
Added check.

drewnoakes commented on src/Aspire.Hosting/Dashboard/DashboardService.cs:
```diff
@@ -0,0 +1,109 @@
+            data.Resources.Add(Resource.FromSnapshot(resource));
+        }
+
+        await responseStream.WriteAsync(new() { InitialData = data }).ConfigureAwait(false);
```
Good observation and I'd be in favour of avoiding the noise. Please file a separate issue for this.

JamesNK replied:
https://github.com/dotnet/aspire/issues/1511

drewnoakes committed:
Throw on unexpected values

drewnoakes committed:
Log when dashboard is disabled

drewnoakes committed:
Remove redundant route on gRPC endpoint

A browser will request using HTTP/1.1, and we only accept HTTP/2, so this route doesn't achieve anything.

drewnoakes committed:
Log reason when dashboard hosting is skipped

drewnoakes committed:
Merge branch 'main' into add-grpc

drewnoakes committed:
Fix build

Removes code made unused in #1378 that triggers build errors.

drewnoakes committed:
Fix errors following merge

drewnoakes committed:
Formatting

drewnoakes committed:
Handle null container ID

drewnoakes committed:
Extract common check

Also reinstate check for non-zero exit code.

drewnoakes committed:
Specify assembly name for eShopLite's app host

The app host exposes this via `Microsoft.Extensions.Hosting.IHostEnvironment.ApplicationName`, which is documented as returning the assembly name. Any `.AppHost` suffix is removed before the name returned.

Because the name of this project is juse `AppHost`, the dashboard for this project would only show `AppHost`, which is misleading.

This sets the assembly name to have the expected naming convention, without requiring the project name to change and look different to sibling projects in eShopLite.

drewnoakes committed:
Ensure the page title displays

With the introduction of gRPC between dashboard and app host, the dashboard can render its UI before the application information request returns.

One negative consequence of this is that the UI's page title and header area do not show the application name. When this data arrives later asynchronously, the UI is not updated.

This change introduces a way for the UI to refresh at a later time, once the application name is available.

drewnoakes committed:
Remove application version from protocol

This was unused, and it's unclear how we'd populate it. We can add it back later if we want it again, but let's not bake it into the protocol just yet.

drewnoakes committed:
Improve parsing of application name

This covers the case that the name does not have a period. For example, one of the sample apps in this repo is named `DaprAppHost`, and the name `Dapr` would be perfect, but the previous code required a period in the suffix (e.g. `.AppHost`).

drewnoakes committed:
Fix potential ANE due to unannotated gRPC code

drewnoakes committed:
Configure logging for dashboard web application

drewnoakes committed:
gRPC channel and client will reconnect themselves

We don't need code to track the state of the channel. This will restore itself and reconnect, according to James.

- Remove explicit connection juggling.
- Construct the channel and client once.
- Dispose the channel when done.

JamesNK commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,355 @@
+    private readonly DashboardService.DashboardServiceClient _client;
+    private Task? _connection;
+
+    public DashboardClient(ILogger<DashboardClient> logger)
+    {
+        _logger = logger;
```
Use logger factory in ctor so it can be passed to GrpcChannel later.
```suggestion
    public DashboardClient(ILoggerFactory loggerFactory)
    {
        _logger = loggerFactory.CreateLogger<DashboardClient>();
```
Then set the logger factory when the `GrpcChannel` is created so its activity is logged by the app: https://learn.microsoft.com/en-us/aspnet/core/grpc/diagnostics?view=aspnetcore-8.0#grpc-client-logging

JamesNK commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,355 @@
+
+            async Task ConnectAsync()
+            {
+                await _channel.ConnectAsync(cancellationToken).ConfigureAwait(false);
+
```
Channel automatically connects on the first call
```suggestion
```

JamesNK commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,355 @@
+                }
+                catch (RpcException ex)
+                {
+                    // Cancellation is reported in an RpcException, so unwrap it to throw the original OperationCanceledException instead.
+                    if (ex.StatusCode == StatusCode.Cancelled && ex.InnerException is OperationCanceledException oce)
+                    {
+                        // Rethrow cancellation
+                        ExceptionDispatchInfo.Capture(oce).Throw();
+                    }
```
You can avoid this by setting ThrowOperationCanceledOnCancellation on the channel. (I really wanted OCE to be the default behavior but was unfortunately overruled)

https://learn.microsoft.com/en-us/aspnet/core/grpc/configuration?view=aspnetcore-8.0#configure-client-options

JamesNK commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,355 @@
+
+                    errorCount++;
+
+                    _logger.LogError("Error {errorCount} watching resources: {error}", errorCount, ex.Message);
```
```suggestion
                    _logger.LogError(ex, "Error {errorCount} watching resources.", errorCount);
```

drewnoakes replied:
This would make a good analyzer.

drewnoakes committed:
Review feedback

drewnoakes commented:
@JamesNK PTAL

JamesNK commented:
I got your branch and ran the app. Project's aren't displaying PID or project file on resources page:

![image](https://github.com/dotnet/aspire/assets/303201/84d9d15f-28c6-4c5d-81f5-279b885f76c8)

JamesNK commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,355 @@
+        }
+    }
+
+    Task IDashboardClient.WhenConnected => _whenConnected.Task;
```
nit: WhenConnected -> Connected.

That's more in line with https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostapplicationlifetime?view=dotnet-plat-ext-8.0

drewnoakes replied:
Those are cancellation tokens. This is more in line with Task.WhenAny etc.

davidfowl replied:
I agree with James on the naming here. If you want another comparison channels has a completion task. Background service has one as well

JamesNK commented on src/Aspire.Dashboard/Components/Layout/MainLayout.razor:
```diff
@@ -9,7 +9,7 @@
         <div class="header-title">
             <FluentAnchor IconStart="@(new AspireIcons.Size32.Logo())" Appearance="Appearance.Stealth" Href="/" Class="logo">
-                @string.Format(Loc[nameof(Layout.MainLayoutDashboardName)], ResourceService.ApplicationName)
+                @DashboardClient.FormatApplicationName(Loc[nameof(Layout.MainLayoutDashboardName)], () => InvokeAsync(StateHasChanged))
```
This could be moved to its own component to hide the work around refresh away. Also, a benefit of it being its own component is the StateHasChanged notification just updates that component rather than the entire page.

drewnoakes replied:
This pattern is also used for `<PageTitle>` on almost every page, and I don't think that can be moved to a component. The refresh is only needed if the page loads before the client is connected, in which case more than just this anchor will be refreshed. I can't see a clean path forward here, but am open to suggestions.

JamesNK replied:
A page title is just a [component with a child render fragment](https://github.com/dotnet/aspnetcore/blob/d461781f80ef098b3d584d6397f3baa1d6b9f285/src/Components/Web/src/Head/PageTitle.cs#L18).

We can:

1. Create a simple component that wraps injecting dashboard client, calling FormatApplication name, and doing formatting.
2. Put the component inside page title, e.g. `<PageTitle><ApplicationName FormatString="Loc[nameof(Layout.MainLayoutDashboardName)]" /><PageTitle>`

drewnoakes replied:
I've got this working locally, but will file a follow up PR so we can review the change independently.

drewnoakes replied:
https://github.com/dotnet/aspire/pull/1555

JamesNK commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,355 @@
+                KeepAlivePingDelay = TimeSpan.FromSeconds(20),
+                KeepAlivePingTimeout = TimeSpan.FromSeconds(10),
+                KeepAlivePingPolicy = HttpKeepAlivePingPolicy.WithActiveRequests,
+                PooledConnectionIdleTimeout = TimeSpan.FromHours(2)
```
What is the reason for the long lifetime? Almost every server will kill an idle connection much faster. It's probably better that the client kills it first to avoid a race between the client starting to use a connection and the server killing it before receiving a request.

drewnoakes replied:
What value would you recommend?

JamesNK replied:
The default is 60 seconds. That's less than Kestrel's idle timeout (130 seconds). Not specifying anything and using the default is fine.

Whatever value we have is probably academic. The watch task is trying to always keep an active request on the connection, so we'll always have one long running connection that never times out.

drewnoakes replied:
I removed this property setter, favouring the default.

JamesNK commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,355 @@
+                                }
+                                else
+                                {
+                                    throw new FormatException($"Unexpected {nameof(WatchResourcesChange)} kind: {change.KindCase}");
```
nit: FormatException isn't the exception I'd expect here. I would expect ArgumentOutOfRangeException or InvalidOperationException

JamesNK commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,355 @@
+                        }
+                        else
+                        {
+                            throw new FormatException($"Unexpected {nameof(WatchResourcesUpdate)} kind: {response.KindCase}");
```
nit: FormatException isn't the exception I'd expect here. I would expect ArgumentOutOfRangeException or InvalidOperationException

drewnoakes replied:
`ArgumentOutOfRangeException` is for arguments, and this is not an argument.

`InvalidOperationException` implies that the requested operation is not applicable given the current state of the object, which does not describe this situation.

The exception here is that the received data is not of the expected format. `FormatException` is used in several scenarios, commonly around parsing, which is what we're doing here.

We could create a custom exception type, but I don't think it's necessary.

JamesNK commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,355 @@
+            return;
+        }
+
+        _connection = Task.Run(() => ConnectAndStayConnectedAsync(_cts.Token), _cts.Token);
+
+        return;
+
+        async Task ConnectAndStayConnectedAsync(CancellationToken cancellationToken)
+        {
+            // Track the number of errors we've seen since the last successfully received message.
+            // As this number climbs, we hold off on 
+            var errorCount = 0;
+
+            await ConnectAsync().ConfigureAwait(false);
```
If the initial gRPC call in `ConnectAsync` fails then the connection task will complete with an error and there will never be resource updates. I think the `GetApplicationInformationAsync` call in `ConnectAsync` needs to move somewhere else. The task should just be for watching resources.

drewnoakes replied:
This method applies to the entire dashboard client, not just gathering resources. Any connection error is grounds to fail. I've refactored the code here to highlight this with clearer method names.

JamesNK replied:
What is the user experience of failure?

For example:

* Customer navigates to the Aspire dashboard hosted by a cloud provider
* Transient error in their `GetApplicationInformationAsync` implementation causes connect task to break

What do they see? From what I understand of the current code, data will never load which isn't a good experience. If they need to refresh their browser to unbreak the dashboard then there needs to be a error UI that tells them that.

drewnoakes replied:
So if `GetApplicationInformationAsync` somehow did throw, the UI would appear with no title and no resources, and with no indication of any problem. I've added a try/catch for this, which at least allows resources to be loaded, and the title remains blank.

JamesNK replied:
Ok. It's not ideal but good enough for now. I think a follow up issue is needed for what happens in the UI if the dashboard service is down.

drewnoakes replied:
https://github.com/dotnet/aspire/issues/1554

drewnoakes committed:
Use concurrent dictionaries

This prevents corruption from concurrent updates of a non-concurrent dictionary.

karolz-ms commented on src/Aspire.Dashboard/Model/IDashboardClientExtensions.cs:
```diff
@@ -0,0 +1,36 @@
+        {
+            // We're not yet connected. Schedule refresh to be invoked
+            // when the connection is established.
+            task.ContinueWith(
```
Do we care whether refresh() is called only once after task completion? I wonder what happens if FormatApplicationName() is called multiple times...

drewnoakes replied:
`FormatApplicationName` is called twice for each page -- once for the `<title>` element and once for the heading shown in the UI. If both require the continuation to trigger refresh, that refresh would happen twice, though in both cases on different parts of the UI, as I understand it. So I don't foresee any issue here.

drewnoakes replied:
I'm going to explore James's suggestion of making a component for this.

drewnoakes replied:
https://github.com/dotnet/aspire/pull/1555

karolz-ms reviewed (approved):
Overall LGTM, just a couple questions

drewnoakes committed:
More spelling words

drewnoakes committed:
Refactor connection code

kvenkatrajan reviewed (approved):
approved

drewnoakes committed:
Fix missing process ID

drewnoakes committed:
Fix resource type checks

drewnoakes commented:
> Projects aren't displaying a PID or project file on resources page

Fixed:

![image](https://github.com/dotnet/aspire/assets/350947/2ccce22b-d00b-4666-9a71-bf1d8c761e8d)

drewnoakes commented on src/Aspire.Dashboard/Model/DashboardClient.cs:
```diff
@@ -0,0 +1,355 @@
+    private ImmutableHashSet<Channel<ResourceViewModelChange>> _outgoingChannels = [];
+    private string? _applicationName;
+
+    private const int StateNone = 0;
```
These are used with `Interlocked` methods that require `int`. If we use an enum we have to cast the values everywhere which is verbose, and the `_state` field must still be an `int`. This is private so doesn't leak out.

drewnoakes committed:
Handle connection exception

JamesNK reviewed (approved):
Some improvements are available to make in comments, but they can be done later. LGTM

davidfowl commented:
Excellent work everyone!
